pkg/sfu/streamallocator.go

```diff
@@ -617,21 +617,20 @@ func (s *StreamAllocator) runStateMachine(event Event) {
 	}
 }
 
-// LK-TODO-START
-// Signal_ADD_TRACK is not useful. Probably can get rid of it.
-// AVAILABLE_LAYERS_ADD/REMOVE should be how track start should
-// be getting an allocation.
-// LK-TODO-END
+// LK-TODO: ADD_TRACK should not reallocate if added track is audio
 func (s *StreamAllocator) runStatePreCommit(event Event) {
 	switch event.Signal {
 	case Signal_ADD_TRACK:
-		// wait for layers add/remove signal
+		s.allocate()
 	case Signal_REMOVE_TRACK:
 		s.allocate()
 	case Signal_ESTIMATE_INCREASE:
-		s.allocate()
+		// should never happen as the intialized capacity is very high
+		s.setState(State_STABLE)
 	case Signal_ESTIMATE_DECREASE:
-		s.allocate()
+		// first estimates could be off as things are ramping up.
+		// Just move to STABLE state and let further streaming drive the state machine
+		s.setState(State_STABLE)
```

Comment:
This was one of the problem spots. I do not know if this is the best. Might be or might need some more thinking.

Basically, we start streaming downstream. For that, we use an initial channel capacity of 100 Mbps (basically something ridiculously large) so that there is no throttling before we know how big the channel is. The first REMB happens 6 - 8 seconds later. With screen share, the REMB is very finicky (as screen share is very sparse). If we do an allocate on getting the first REMB, there is a chance that the screen share is stalled. This is because, we take the current bit rate when doing allocation.

This is one of the significant challenges of `streamallocator`, i. e. time alignment. REMB was potentially measured a while back. But, we use that number to allocate based on current bitrate. So, if the intervening period if there was movement on the screen, the bitrate could have spiked and the REMB value is probably stale.

A few things to try
- Try this change (i. e. change to STABLE state and further streaming drive REMB and move the state machine in whichever direction it needs to go)
- Forward padding packets also (but don't want to do this as somebody joining in the middle will fire a ton of padding packets and that might disrupt existing flows)
- Do probing downstream in the first few seconds in addition to forwarding any streams and let us see how the channel behaves
- Measure up stream bitrates over a much longer window so that we really get some steady state bitrate (but even that could be jumpy with screen share)

A good challenge. Basically, we will have to look at how much we can avoid re-allocation of streams. Any re-allocation means potential for REMB numbers and current state of up streams going not being in sync causing trouble. Need another thorough pass to figure how many places can we avoid re-allocation.

Reply:
thanks for such detailed explanation. would be great to leave this in the code as well in streamallocator so it doesn't get lost with the PR.

Reply:
Will add it inline in code.